File: design/lru_cache.py
"""
LeetCode 146 - LRU Cache (Medium)
https://leetcode.com/problems/lru-cache/

Design a data structure that follows the constraints of a Least
Recently Used (LRU) cache.

Implement the LRUCache class:
  - LRUCache(int capacity)        Initializes the LRU cache with positive
                                   size capacity.
  - int get(int key)              Returns the value of the key if it
                                   exists, otherwise returns -1.
  - void put(int key, int value)  Updates the value of the key if it
                                   exists. Otherwise, adds the key-value
                                   pair. If the number of keys exceeds
                                   capacity, evict the least recently
                                   used key.

Both get and put must run in O(1) average time complexity.

Solution idea:
  Doubly-linked list (most-recently-used at head, least at tail) plus a
  dict mapping key -> node for O(1) lookup. `get` finds the node, splices
  it out, and re-inserts it at the head. `put` either replaces an
  existing node (move-to-head semantics again) or appends a new one,
  evicting the tail neighbour when over capacity.

Typing note:
  Each `Node` initialises `_next` / `_prev` to `self` (self-reference),
  so the field type stays `Node` (not `Optional[Node]`) without any
  infinite recursion. LinkedList wires up real prev/next pointers
  immediately after construction. Sentinel head and tail keep their
  outward pointers self-looped (head.prev == head, tail.next == tail),
  which keeps every accessor's return type as `Node` and avoids the
  pyre cascade you hit with `Optional[Node]` (where
  `prev_node.prev = node` complains that prev_node could be None).
"""
from __future__ import annotations


class Node:
    def __init__(self, key: int, val: int) -> None:
        self.__val = val
        self.__key = key
        # Self-reference avoids the `Optional[Node]` cascade in pyre
        # while never triggering infinite recursion at construction
        # (a name binding is not a constructor call). The owning
        # LinkedList rewires real prev/next pointers right away.
        self._next: Node = self
        self._prev: Node = self

    @property
    def next(self) -> Node:
        return self._next

    @next.setter
    def next(self, node: Node) -> None:
        self._next = node

    @property
    def prev(self) -> Node:
        return self._prev

    @prev.setter
    def prev(self, node: Node) -> None:
        self._prev = node

    @property
    def key(self) -> int:
        return self.__key

    @property
    def val(self) -> int:
        return self.__val


class LinkedList:
    def __init__(self) -> None:
        self.head: Node = Node(-1, -1)
        self.tail: Node = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def add_node_next_to_head(self, node: Node) -> None:
        prev_node = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = prev_node
        prev_node.prev = node

    def delete_node(self, node: Node) -> None:
        next_node = node.next
        prev_node = node.prev
        next_node.prev = prev_node
        prev_node.next = next_node

    def delete_last_node(self) -> None:
        last_node = self.tail.prev
        if last_node is self.head:
            return
        else:
            self.delete_node(last_node)
        

class LRUCache(object):

    def __init__(self, capacity: int) -> None:
        """
        :type capacity: int
        """
        self.lru_list = LinkedList()
        self.capacity = capacity
        self.lru: dict[int, Node] = {}
        

    def get(self, key: int) -> int:
        """
        :type key: int
        :rtype: int
        """
        # print(self.lru_list)
        if key not in self.lru:
            return -1
        else:
            node = self.lru[key]
            self.lru_list.delete_node(node)
            self.lru_list.add_node_next_to_head(node)
            return node.val
        

    def put(self, key: int, value: int) -> None:
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        # print(self.lru_list)
        new_node = Node(key, value)
        self.lru_list.add_node_next_to_head(new_node)
        if key in self.lru:
            old_node = self.lru[key]
            self.lru_list.delete_node(old_node)
            self.lru[key] = new_node
        else:
            if len(self.lru) >= self.capacity:
                last_node = self.lru_list.tail.prev
                del self.lru[last_node.key]
                self.lru_list.delete_node(last_node)
            self.lru[key] = new_node
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)


if __name__ == "__main__":
    # LeetCode example 1
    c = LRUCache(2)
    c.put(1, 1)                     # cache = {1=1}
    c.put(2, 2)                     # cache = {1=1, 2=2}
    assert c.get(1) == 1            # touches 1; LRU order is now {2, 1}
    c.put(3, 3)                     # evicts 2; cache = {1=1, 3=3}
    assert c.get(2) == -1           # 2 was evicted
    c.put(4, 4)                     # evicts 1; cache = {3=3, 4=4}
    assert c.get(1) == -1           # 1 was evicted
    assert c.get(3) == 3
    assert c.get(4) == 4

    print("lru_cache.py: all tests passed")

File: c++/design/lru_cache.cpp
// LeetCode 146 - LRU Cache (Medium)
// https://leetcode.com/problems/lru-cache/
//
// Design a data structure that follows the constraints of a Least
// Recently Used (LRU) cache.
//
// Implement the LRUCache class:
//   - LRUCache(int capacity)        Initialize with positive capacity.
//   - int get(int key)              Return the value of the key if
//                                    present, else -1.
//   - void put(int key, int value)  Update or insert (key, value).
//                                    Evict the least recently used key
//                                    when over capacity.
//
// Both get and put must run in O(1) average time complexity.
//
// Solution idea:
//   Doubly-linked list of nodes ordered most-recently-used at the
//   front, plus an unordered_map<int, unique_ptr<Node>> for O(1)
//   key -> node lookup. Sentinel head_/tail_ nodes simplify boundary
//   handling: every real node always has non-null prev and next.
//   `move_to_front` is detach + insert_after_head and is the work
//   shared by `get` and the "key already exists" branch of `put`.
//   Ownership lives on the map (unique_ptr); the linked list stores
//   raw pointers for navigation, so erasing a key from the map
//   destroys the node and all pointers to it disappear together.

#include <cstddef>
#include <memory>
#include <unordered_map>

class LRUCache {
private:
    struct Node {
        int key = 0;
        int value = 0;

        Node* prev = nullptr;
        Node* next = nullptr;

        Node() = default;

        Node(int k, int v)
            : key(k), value(v) {}
    };

public:
    explicit LRUCache(int capacity)
        : capacity_(capacity > 0 ? static_cast<std::size_t>(capacity) : 0)
    {
        head_.next = &tail_;
        tail_.prev = &head_;
    }

    int get(int key) {
        auto it = nodes_.find(key);

        if (it == nodes_.end()) {
            return -1;
        }

        Node* node = it->second.get();
        move_to_front(node);

        return node->value;
    }

    void put(int key, int value) {
        if (capacity_ == 0) {
            return;
        }

        auto it = nodes_.find(key);

        if (it != nodes_.end()) {
            Node* node = it->second.get();
            node->value = value;
            move_to_front(node);
            return;
        }

        if (nodes_.size() == capacity_) {
            Node* victim = tail_.prev;
            detach(victim);
            nodes_.erase(victim->key);
        }

        auto new_node = std::make_unique<Node>(key, value);
        Node* raw_node = new_node.get();

        insert_after_head(raw_node);
        nodes_.emplace(key, std::move(new_node));
    }

private:
    std::size_t capacity_;

    Node head_;
    Node tail_;

    std::unordered_map<int, std::unique_ptr<Node>> nodes_;

    void detach(Node* node) {
        Node* before = node->prev;
        Node* after = node->next;

        before->next = after;
        after->prev = before;

        node->prev = nullptr;
        node->next = nullptr;
    }

    void insert_after_head(Node* node) {
        Node* old_first = head_.next;

        node->prev = &head_;
        node->next = old_first;
        
        head_.next = node;
        old_first->prev = node;
    }

    void move_to_front(Node* node) {
        detach(node);
        insert_after_head(node);
    }
};


// ----- local tests (not part of the LeetCode submission) -----
// Build and run:
//   g++ -std=c++17 lru_cache.cpp -o /tmp/lru && /tmp/lru
#include <cassert>
#include <iostream>

int main() {
    // LeetCode example 1
    LRUCache c(2);
    c.put(1, 1);                    // cache = {1=1}
    c.put(2, 2);                    // cache = {1=1, 2=2}
    assert(c.get(1) == 1);          // touches 1; LRU order now {2, 1}
    c.put(3, 3);                    // evicts 2; cache = {1=1, 3=3}
    assert(c.get(2) == -1);         // 2 was evicted
    c.put(4, 4);                    // evicts 1; cache = {3=3, 4=4}
    assert(c.get(1) == -1);         // 1 was evicted
    assert(c.get(3) == 3);
    assert(c.get(4) == 4);

    std::cout << "lru_cache.cpp: all tests passed" << std::endl;
    return 0;
}